javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll

```diff
@@ -108,3 +117,317 @@ private class NosqlInjectionSinkCharacteristic extends EndpointCharacteristic {
     confidence = 1.0
   }
 }
+
+/*
+ * Characteristics that are indicative of not being a sink of any type.
+ */
+
+/**
+ * A characteristic that is an indicator of not being a sink of any type, because it's an argument that has a manual
+ * model.
+ */
+abstract private class OtherModeledArgumentCharacteristic extends EndpointCharacteristic {
+  bindingset[this]
+  OtherModeledArgumentCharacteristic() { any() }
+}
+
+/**
+ * A characteristic that is an indicator of not being a sink of any type, because it's an argument to a function of a
+ * builtin object.
+ */
+abstract private class ArgumentToBuiltinFunctionCharacteristic extends OtherModeledArgumentCharacteristic {
+  bindingset[this]
+  ArgumentToBuiltinFunctionCharacteristic() { any() }
+}
+
+/**
+ * A high-confidence characteristic that indicates that an endpoint is not a sink of any type.
+ */
+abstract private class NotASinkCharacteristic extends OtherModeledArgumentCharacteristic {
```

Comment:
> > This class reads a bit strange and I think this is worth hashing out.
> > This abstract class wants to state that any of the 20 classes extending it will have a `getImplications` predicate that states that any of the endpoints satisfying their `getEndpoints` predicate are a high-confidence-isPositive-`Negatives`.
> > I'm surprised that the class does NOT state that these endpoints are also:
> > 
> > * high-confidence-_isNegative_-NosqlInjectionSinkType
> > * high-confidence-_isNegative_-SqlInjectionSinkType
> > * high-confidence-_isNegative_-XssSinkType
> > * high-confidence-_isNegative_-TaintedPathSinkType
> 
> That would be redundant. If we already know with high confidence that these endpoints are not sinks of _any_ type, why would we need to specify also that they are not NoSQL sinks, not SQL sinks, etc.? Also, that would require adding every new sink type we want to boost to this list.
> 
> The reason we have the ability to specify something like `high-confidence-_isNegative_-NosqlInjectionSinkType` is for characteristics that are inherently sink-type specific, such as our current query-specific endpoint filters. If I have good reason to say a certain set of endpoints are not NoSQL sinks, but I don't know anything about whether or not they're, say, Xss sinks, then I'll use `high-confidence-_isNegative_-NosqlInjectionSinkType`.
> 
> > Or maybe it should state instead that these endpoints are:
> > 
> > * _low-confidence_-isPositive-NosqlInjectionSinkType
> > * _low-confidence_-isPositive-SqlInjectionSinkType
> > * _low-confidence_-isPositive-XssSinkType
> > * _low-confidence_-isPositive-TaintedPathSinkType
> 
> This would be wrong. `_low-confidence_-isPositive-NosqlInjectionSinkType` means that I think this endpoint _is_ a NoSQL sink, but I'm not very confident about that assertion. See [the definition of isPositiveIndicator](https://github.com/github/codeql/blob/ac14b6d68549e62571134e70c334bef82f826ffc/javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll#L36): `isPositiveIndicator: If true, this characteristic indicates that this endpoint _is_ a member of the class; if false, it indicates that it _isn't_ a member of the class.`
> 
> Low-confidence indicators may come in handy for speeding up inference: If we think an endpoint is not very likely to be a sink, it may not be worth featurizing and scoring it. For the training set, OTOH, we probably won't look at characteristics that give only a weak indication, because we want to be fairly certain that most of our training samples have a correct label.
> 
> > Or should it state both?
> > I'm a bit puzzled.. I don't have a good intuition about this confidence mechanism; that there's a boolean IN ADDITION to a confidence score seems to add to my confusion (perhaps replacing that by a confidence score [-1, 1] would help some?).
> 
> See the [definition of the confidence](https://github.com/github/codeql/blob/ac14b6d68549e62571134e70c334bef82f826ffc/javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll#L38), and especially the example about what a low-confidence implication tells us: `A confidence near zero means this characteristic is a very weak indicator of whether or not the endpoint belongs to the class.` In practice we'll probably never bother writing EndpointCharacteristics that have confidence very close to zero, because characteristics that tell us almost nothing aren't useful. As mentioned above, though, we might use characteristics with confidence well below 1 to filter out unlikely candidates at inference time, to speed up code scanning, at the risk of maybe missing out on a few sinks our models might have otherwise detected.
> 
> Is there anything I could do to clarify the doc string more? If this was confusing for you, it might confuse others down the line as well.
> 
> Also, if it's still confusing, maybe we should sync about it 🙂

Reply:
The permalinks are now fixed -- sorry about that!

Reply:
After our sync discussion yesterday, I'm satisfied with the direction this is going in. I'm not 100% convinced that the first formulation is really redundant **but now is not the best time to figure that out - time will tell**.

If we'd have found out that any of these additional relations would've been necessary, we'd have needed to take a step back, and I'm glad we didn't :)

I'm resolving this conversation.